File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>&h){
        int l=0;
        long long r=h.size()-1;
        int lm=0,rm=0,res=0;
        while(l<r){
            if(h[l]<h[r]){
                if(h[l]>=lm){
                    lm=h[l];
                }
                else{
                    res+=lm-h[l];
                } 
                l++;
            }else{
                if(h[r]>=rm){
                    rm=h[r];
                }
                else{
                    res+=rm-h[r];
                }
                r--;
            }
        }
        return res;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l=0
        r=len(height)-1
        lm=0
        rm=0
        ans=0
        while l<=r:
            if height[l]<height[r]:
                if height[l]>=lm:
                    lm=height[l]
                else:
                    ans+=lm-height[l]
                l+=1
            else:
                if height[r]>=rm:
                    rm=height[r]
                else:
                    ans+=rm-height[r]
                r-=1
        return ans
